bolt-modules/boltlib/lib/puppet/functions/vars.rb

```diff
@@ -0,0 +1,20 @@
+require 'bolt/error'
+
+Puppet::Functions.create_function(:vars) do
+  dispatch :vars do
+    param 'Target', :target
+    return_type 'Hash[String, Data]'
+  end
+
+  def vars(target)
+    unless Puppet[:tasks]
+      raise Puppet::ParseErrorWithIssue.from_issue_and_stack(
+        Puppet::Pops::Issues::TASK_OPERATION_NOT_SUPPORTED_WHEN_COMPILING, operation: 'set_var'
+      )
+    end
+
+    inventory = Puppet.lookup(:bolt_inventory) { nil }
+
+    inventory.vars(target)
```

Comment:
This will return a mutable hash which is a no-go. It can be resolved in one of three ways.
1. by having set_var perform a "copy on write", i.e. it modifies a cloned hash for the given target. By doing so, anyone holding on to the old hash will not get affected by the set_var call. The target hash should then always be frozen.
2. by letting vars perform a "copy on read", i.e. it returns a copy of the hash which should be frozen.
3. by removing the vars method and instead provide iterative methods to access its entries (as I suggested earlier).

Reply:
We already return mutable hashes/object all over the place. If we decide to change that it probably needs to be it's own effort.

Reply:
The hash in itself may be mutable as long as it is a private copy that cannot be changed by ruby code. This one is intended to be changed for each invocation of set_var. In essence, that's the same thing as introducing a mutable object to in Puppet. We can't do that.

Reply:
We've already done that with Result objects.

Reply:
In that case that was a mistake and the Result objects need to change. It doesn't mean that introducing new similar mistakes is acceptable.

Reply:
where does this happen with the `Result` object? I can't find it.

Reply:
Nothing is frozen about the value hash before it's returned.

Reply:
No, it's not frozen, but it very well could be (and it's a good idea to ensure that it is). There's no way it can be manipulated unless someone would change the error_hash from Ruby code. AFAICS, no one does.